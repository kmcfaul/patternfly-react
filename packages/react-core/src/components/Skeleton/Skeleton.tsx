import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/Skeleton/skeleton';
import { css } from '@patternfly/react-styles';

export interface SkeletonProps extends React.HTMLProps<HTMLDivElement> {
  /** Additional classes added to the Skeleton */
  className?: string;
  /** The percentage width of the Skeleton */
  width?: string;
  /** The percentage height of the Skeleton */
  height?: string;
  /** The font size height of the Skeleton */
  fontSize?: 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl';
  /** The shape of the Skeleton */
  shape?: 'circle' | 'square' | 'rectangle';
  /** Width of the shape */
  shapeWidth?: 'sm' | 'md' | 'lg';
  /** Height of the shape */
  shapeHeight?: 'sm' | 'md' | 'lg';
}

const shapeWidths = {
  sm: styles.modifiers.widthSm,
  md: styles.modifiers.widthMd,
  lg: styles.modifiers.widthLg
};

const shapeHeights = {
  sm: styles.modifiers.heightSm,
  md: styles.modifiers.heightMd,
  lg: styles.modifiers.heightLg
};

export const Skeleton: React.FunctionComponent<SkeletonProps> = ({
  className = '',
  width = '',
  height = '',
  fontSize,
  shape,
  shapeWidth,
  shapeHeight,
  ...props
}: SkeletonProps) => {
  const presetWidth = Object.values(styles.modifiers).find(key => key === `pf-m-width-${width}`);
  const presetHeight = Object.values(styles.modifiers).find(key => key === `pf-m-height-${height}`);
  const fontHeight = Object.values(styles.modifiers).find(key => key === `pf-m-text-${fontSize}`);
  return (
    <div
      {...props}
      className={css(
        styles.skeleton,
        presetWidth,
        presetHeight,
        fontSize && fontHeight,
        shape === 'circle' && styles.modifiers.circle,
        shape === 'square' && styles.modifiers.square,
        shapeWidth && shapeWidths[shapeWidth],
        shapeHeight && shapeHeights[shapeHeight],
        className
      )}
      {...(((width && !presetWidth) || (height && !presetHeight)) && {
        style: {
          width: width && !presetWidth ? `${width}%` : undefined,
          height: height && !presetHeight ? `${height}%` : undefined,
          ...props.style
        }
      })}
    ></div>
  );
};
Skeleton.displayName = 'Skeleton';
