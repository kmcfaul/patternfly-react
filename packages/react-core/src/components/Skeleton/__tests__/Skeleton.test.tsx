import React from 'react';
import { shallow } from 'enzyme';
import { Skeleton } from '../Skeleton';

describe('skeleton', () => {
  test('default', () => {
    const view = shallow(<Skeleton />);
    expect(view).toMatchSnapshot();
  });

  ['25', '33', '50', '66', '75'].forEach((width: string) =>
    test(`skeleton with ${width} width`, () => {
      const view = shallow(<Skeleton width={width} />);
      expect(view).toMatchSnapshot();
    })
  );

  ['25', '33', '50', '66', '75', '100'].forEach((height: string) =>
    test(`skeleton with ${height} height`, () => {
      const view = shallow(<Skeleton height={height} />);
      expect(view).toMatchSnapshot();
    })
  );

  ['sm', 'md', 'lg', 'xl', '2xl', '3xl', '4xl'].forEach((fontSize: string) =>
    test(`skeleton with ${fontSize} font size`, () => {
      const view = shallow(<Skeleton fontSize={fontSize as 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl'} />);
      expect(view).toMatchSnapshot();
    })
  );

  test('circle skeleton', () => {
    const view = shallow(<Skeleton shape="circle" />);
    expect(view).toMatchSnapshot();
  });

  test('square skeleton', () => {
    const view = shallow(<Skeleton shape="square" />);
    expect(view).toMatchSnapshot();
  });

  test('rectangle skeleton', () => {
    const view = shallow(<Skeleton shapeHeight="sm" shapeWidth="md" />);
    expect(view).toMatchSnapshot();
    const viewMd = shallow(<Skeleton shapeHeight="md" shapeWidth="lg" />);
    expect(viewMd).toMatchSnapshot();
    const viewLg = shallow(<Skeleton shapeHeight="lg" />);
    expect(viewLg).toMatchSnapshot();
  });

  ['sm', 'md', 'lg'].forEach((width: string) =>
    test(`shape skeleton with ${width} width`, () => {
      const view = shallow(<Skeleton shape="circle" shapeWidth={width as 'sm' | 'md' | 'lg'} />);
      expect(view).toMatchSnapshot();
      const viewSquare = shallow(<Skeleton shape="square" shapeWidth={width as 'sm' | 'md' | 'lg'} />);
      expect(viewSquare).toMatchSnapshot();
    })
  );
});
