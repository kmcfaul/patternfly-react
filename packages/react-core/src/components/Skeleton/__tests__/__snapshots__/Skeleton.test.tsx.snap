// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skeleton circle skeleton 1`] = `
<div
  className="pf-c-skeleton pf-m-circle"
/>
`;

exports[`skeleton default 1`] = `
<div
  className="pf-c-skeleton"
/>
`;

exports[`skeleton rectangle skeleton 1`] = `
<div
  className="pf-c-skeleton pf-m-width-md pf-m-height-sm"
/>
`;

exports[`skeleton rectangle skeleton 2`] = `
<div
  className="pf-c-skeleton pf-m-width-lg pf-m-height-md"
/>
`;

exports[`skeleton rectangle skeleton 3`] = `
<div
  className="pf-c-skeleton pf-m-height-lg"
/>
`;

exports[`skeleton shape skeleton with lg width 1`] = `
<div
  className="pf-c-skeleton pf-m-circle pf-m-width-lg"
/>
`;

exports[`skeleton shape skeleton with lg width 2`] = `
<div
  className="pf-c-skeleton pf-m-square pf-m-width-lg"
/>
`;

exports[`skeleton shape skeleton with md width 1`] = `
<div
  className="pf-c-skeleton pf-m-circle pf-m-width-md"
/>
`;

exports[`skeleton shape skeleton with md width 2`] = `
<div
  className="pf-c-skeleton pf-m-square pf-m-width-md"
/>
`;

exports[`skeleton shape skeleton with sm width 1`] = `
<div
  className="pf-c-skeleton pf-m-circle pf-m-width-sm"
/>
`;

exports[`skeleton shape skeleton with sm width 2`] = `
<div
  className="pf-c-skeleton pf-m-square pf-m-width-sm"
/>
`;

exports[`skeleton skeleton with 2xl font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-2xl"
/>
`;

exports[`skeleton skeleton with 3xl font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-3xl"
/>
`;

exports[`skeleton skeleton with 4xl font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-4xl"
/>
`;

exports[`skeleton skeleton with 25 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-25"
/>
`;

exports[`skeleton skeleton with 25 width 1`] = `
<div
  className="pf-c-skeleton pf-m-width-25"
/>
`;

exports[`skeleton skeleton with 33 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-33"
/>
`;

exports[`skeleton skeleton with 33 width 1`] = `
<div
  className="pf-c-skeleton pf-m-width-33"
/>
`;

exports[`skeleton skeleton with 50 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-50"
/>
`;

exports[`skeleton skeleton with 50 width 1`] = `
<div
  className="pf-c-skeleton pf-m-width-50"
/>
`;

exports[`skeleton skeleton with 66 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-66"
/>
`;

exports[`skeleton skeleton with 66 width 1`] = `
<div
  className="pf-c-skeleton pf-m-width-66"
/>
`;

exports[`skeleton skeleton with 75 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-75"
/>
`;

exports[`skeleton skeleton with 75 width 1`] = `
<div
  className="pf-c-skeleton pf-m-width-75"
/>
`;

exports[`skeleton skeleton with 100 height 1`] = `
<div
  className="pf-c-skeleton pf-m-height-100"
/>
`;

exports[`skeleton skeleton with lg font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-lg"
/>
`;

exports[`skeleton skeleton with md font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-md"
/>
`;

exports[`skeleton skeleton with sm font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-sm"
/>
`;

exports[`skeleton skeleton with xl font size 1`] = `
<div
  className="pf-c-skeleton pf-m-text-xl"
/>
`;

exports[`skeleton square skeleton 1`] = `
<div
  className="pf-c-skeleton pf-m-square"
/>
`;
